tiatoolbox/models/engine/io_config.py

```diff
@@ -0,0 +1,407 @@
+
+
+@dataclass
+class ModelIOConfigABC:
+    """Defines a data class for holding a deep learning model's I/O information.
+
+    Enforcing such that following attributes must always be defined by
+    the subclass.
+
+    Args:
+        input_resolutions (list(dict)):
+            Resolution of each input head of model inference, must be in
+            the same order as `target model.forward()`.
+        patch_input_shape (:class:`numpy.ndarray`, list(int), tuple(int, int)):
+            Shape of the largest input in (height, width).
+        stride_shape (:class:`numpy.ndarray`, list(int), tuple(int)):
+            Stride in (x, y) direction for patch extraction.
+        output_resolutions (list(dict)):
+            Resolution of each output head from model inference, must be
+            in the same order as target model.infer_batch().
+
+    Attributes:
+        input_resolutions (list(dict)):
+            Resolution of each input head of model inference, must be in
+            the same order as `target model.forward()`.
+        patch_input_shape (:class:`numpy.ndarray`, list(int), tuple(int, int)):
+            Shape of the largest input in (height, width).
+        stride_shape (:class:`numpy.ndarray`, list(int), tuple(int, int)):
+            Stride in (x, y) direction for patch extraction.
+        output_resolutions (list(dict)):
+            Resolution of each output head from model inference, must be
+            in the same order as target model.infer_batch().
+        highest_input_resolution (dict):
+            Highest resolution to process the image based on input and
+            output resolutions. This helps to read the image at the optimal
+            resolution and improves performance.
+
+    """
+
+    input_resolutions: List[dict]
+    patch_input_shape: Union[List[int], np.ndarray, Tuple[int, int]]
+    stride_shape: Union[List[int], np.ndarray, Tuple[int, int]] = None
+    output_resolutions: List[dict] = field(default_factory=list)
+
+    def __post_init__(self):
+        if self.stride_shape is None:
+            self.stride_shape = self.patch_input_shape
+
+        self.resolution_unit = self.input_resolutions[0]["units"]
+
+        if self.resolution_unit == "mpp":
+            self.highest_input_resolution = min(
+                self.input_resolutions, key=lambda x: x["resolution"]
+            )
+        else:
+            self.highest_input_resolution = max(
+                self.input_resolutions, key=lambda x: x["resolution"]
+            )
+
+        self._validate()
+
+    def _validate(self):
+        """Validate the data format."""
+        resolutions = self.input_resolutions + self.output_resolutions
+        units = [v["units"] for v in resolutions]
+        units = np.unique(units)
+
+        if len(units) != 1:
+            raise ValueError(
+                f"Invalid resolution units `{units}`. "
+                f"The resolution units must be unique."
```

Comment:
```suggestion
                f"Mixing resolution units is not allowed."
```

Reply:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/618/commits/cd4acbb590f4021399117ccf9d90ff3963225818